### PR TITLE
Display screenshots and draw a border around preselected

### LINKED ISSUE #2: Draw a rectangle that can display ZPixmaps
Use xcb to create a window with customizable background and draw screenshot of current desktop there with a border and caption.
- Screenshot should be made at program start (in the next version screenshots will be fetched from daemon)
- Caption should be equal to `_NET_DESKTOP_NAMES`

### PR DESCRIPTION
**Pull request by mangalinor**
1. Working GUI that displays desktop screenshots received from daemon.
2. Colored border around current desktop
3. Most parameters can be changed in `config.hpp`.

### CODE REVIEW

File: src/config.hpp [hunk 1]
```diff
@@ -4,3 +4,15 @@ const int k_dexpo_width = 500;
...
+// Highlight color
+// Red (0xFFFF0000) recommended
+const int k_dexpo_hcolor = 0xFFFF0000;
+// Higlight border width
+// Should not be less than 3 or more than 8
+// Otherwise it will be poorly visible or buggy
+const int k_dexpo_hwidth = 3;
+// Background color
+// White (0xFFFFFFFF) recommended
+const int k_dexpo_bcolor = 0xFFFFFFFF;
```
Comment (mmskv) [inline src/config.hpp]: Add a newline at the end of file

File: src/window.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+#include <vector>
+#include <xcb/xproto.h>
+
+/* NOTE: First part is mostly close to DesktopPixmap */
+window::window (const int16_t x,         ///< x coordinate of the top left corner
+                const int16_t y,         ///< y coordinate of the top left corner
+                const uint16_t width,    ///< width of display
+                const uint16_t height,   ///< height of display
+                const std::string &name, ///< _NET_WM_NAME of display
+                const uint32_t parent)
```
Comment (mmskv) [inline src/window.cpp]: Remove unused variables

File: src/dexpo.cpp [hunk 3]
```diff
@@ -1,5 +1,95 @@
...
+      conf_height += k_dexpo_padding;   // Add the padding before first screenshot
+      conf_width += 2* k_dexpo_padding; // Add the paddng at both sizes of interface
+      for (const auto &dexpo_pixmap : v)
+        {
+          conf_height += dexpo_pixmap.height;
+          conf_height += k_dexpo_padding;
+        }
+    };
+
+  window w (0, 0, conf_width, conf_height, "Win_1", XCB_NONE);
```
Comment (mmskv) [inline src/dexpo.cpp]: Get window `x` and `y` from `config.hpp`

File: src/window.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    : drawable (x, y, width, height)
+{
+  this->name = name;
+  this->b_width = 0;
```
Comment (mmskv) [inline src/window.cpp]: Get window border color and size from config

File: src/window.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  // borders
+  uint32_t mask = 0;
+  uint32_t values[3];
+  const bool override_redirect[] = {
+    true
+  };
+  mask = XCB_CW_BACK_PIXEL | XCB_CW_EVENT_MASK | XCB_CONFIG_WINDOW_STACK_MODE;
+  values[0] = k_dexpo_bcolor;                               // Background color
+  values[1] = XCB_EVENT_MASK_EXPOSURE | XCB_EVENT_MASK_KEY_PRESS; // Will be used in future to handle events
+  values[2] = XCB_STACK_MODE_ABOVE;                               // Places created window on top
```
Comment (mmskv) [inline src/window.cpp]: Max line width is 80 chars

File: src/window.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    {
+      x = get_screen_position (DesktopNumber);
+    }
+  
+  // We can't control rectangle's thickness, so instead we spawn several close to each other
+  xcb_rectangle_t rectangles[] = {
+    {x-k_dexpo_hwidth, y-k_dexpo_hwidth, k_dexpo_hwidth, storage[DesktopNumber].height+2*k_dexpo_hwidth},                       // left border
+    {x+storage[DesktopNumber].width, y-k_dexpo_hwidth, k_dexpo_hwidth, storage[DesktopNumber].height+2*k_dexpo_hwidth},  // right border
+    {x-1, y-k_dexpo_hwidth, storage[DesktopNumber].width+1, k_dexpo_hwidth},                        // top border
+    {x-1, y+storage[DesktopNumber].height, storage[DesktopNumber].width+1, k_dexpo_hwidth}   // bottom border
```
Comment (mmskv) [inline src/window.cpp]: Unreadable

File: src/window.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  // Drawing the highlighting itself
+  xcb_poly_fill_rectangle(window::c_, this->id, desktop_pixmap::gc_, 4, rectangles);
+
+  // Changing color
+  uint32_t mask = XCB_GC_FOREGROUND;
+  values[0] = k_dexpo_hcolor;
+  xcb_change_gc (c_, desktop_pixmap::gc_, mask, &values);
+}
+
+void window::remove_highlight (int DesktopNumber)
```
Comment (mmskv) [inline src/window.cpp]: Change `DesktopNumber` case to snake_case

File: src/window.hpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ */
+class window : public drawable
+{
+public:
+  inline static xcb_gcontext_t gc_; ///< Graphic context
+  uint32_t id;                      // Identificator for Window
+  short b_width;    // Border width NOTE: Maybe it could be changed to constant
+  std::string name; // _NET_WM_NAME of display, MAYBE UNNEEDED THERE
+  uint32_t parent;  // The window in which GUI will spawn, mostly root
+  std::vector <dexpo_pixmap> storage;
```
Comment (mmskv) [inline src/window.hpp]: Rename variable

File: src/window.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+  // Connection to X-server, as well as generating id before drawing window
+  // itself
+  // Sets default parent to root
+  if (this->parent == XCB_NONE)
+    {
+      this->id = xcb_generate_id (drawable::c_);
+      this->parent = drawable::screen_->root;
+      create_window ();
+    }
```
Comment (mmskv) [inline src/window.cpp]: This is redundant. Remove `this->parent` and explicitly use `drawable::screen_->root` or add `this->root` variable

File: src/window.cpp (hunk 7 above)
Comment (mmskv) [reply src/window.cpp]: Don't write the same function twice.
Just use `window::highlight_window` and pass it a highlight color.
This way you can pass it `k_dexpo_hcolor` when you want to highlight the window and
`k_dexpo_bcolor` when you want to remove the highlight

**Review by mmskv — CHANGES_REQUESTED** [review]
**Fucking use clang-tidy and clang-format**

1. Add a keyboard event handler
  - Kill window on `<Esc>`
  - Move highlight between desktops with arrow keys and `hjkl`

File: src/window.cpp [hunk 10]
```diff
@@ -26,15 +26,14 @@ window::create_window ()
...
   mask = XCB_CW_BACK_PIXEL | XCB_CW_EVENT_MASK | XCB_CONFIG_WINDOW_STACK_MODE;
-  values[0] = dexpo_bgcolor;                               // Background color
-  values[1] = XCB_EVENT_MASK_EXPOSURE | 
-  XCB_EVENT_MASK_KEY_PRESS; // Will be used in future to handle events
-  values[2] = XCB_STACK_MODE_ABOVE;                               // Places created window on top
-  xcb_create_window (window::c_,                    /* Connection, separate from one of daemon */
+  values[0] = dexpo_bgcolor; // Background color
+  values[1]
+      = XCB_EVENT_MASK_EXPOSURE
+        | XCB_EVENT_MASK_KEY_PRESS; // Will be used in future to handle events
```
Comment (mmskv) [inline src/window.cpp]: Move comments so expressions do not get stretched to multiple lines

File: src/window.hpp [hunk 11]
```diff
@@ -15,16 +15,16 @@ class window : public drawable
 public:
   inline static xcb_gcontext_t gc_; ///< Graphic context
   uint32_t id;                      // Identificator for Window
-  short b_width;    // Border width NOTE: Maybe it could be changed to constant
-  std::vector <dexpo_pixmap> storage;
+  short b_width; // Border width NOTE: Maybe it could be changed to constant
+  std::vector<dexpo_pixmap> storage;
```
Comment (mmskv) [inline src/window.hpp]: Rename to something more fitting. Like `pixmaps` or `desktops` or `screenshots`

**Review by mmskv — CHANGES_REQUESTED** [review]
lgtm

**Review by mmskv — APPROVED** [review]

File: src/dexpo.cpp [hunk 12]
```diff
@@ -1,5 +1,99 @@
...
+          conf_height += dexpo_padding;
+        }
+    };
+
+  window w (dexpo_x, dexpo_y, conf_width, conf_height);
+  w.pixmaps = v;
+  // Mapping pixmap onto window
+  while (1)
+    {
+      usleep (1000);
```
Comment (TheHussky) [inline src/dexpo.cpp:66]: so it's sleep in daemon and usleep here?

File: src/window.cpp [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      = { // left border
+          {
+              int16_t (x - dexpo_hlwidth),          /* x */
+              int16_t (y - dexpo_hlwidth),          /* y */
+              uint16_t (dexpo_hlwidth),             /* width */
+              uint16_t (height + 2 * dexpo_hlwidth) /* height */
+          },
+          // right border
+          {
+              int16_t (x + width), /* x */
```
Comment (TheHussky) [inline src/window.cpp]: format?

File: src/window.cpp [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+          {
+              int16_t (x - 1),             /* x */
+              int16_t (y - dexpo_hlwidth), /* y */
+              uint16_t (width + 1),        /* width */
+              uint16_t (dexpo_hlwidth)     /* height */
+          },
+          // bottom border
+          {
+              int16_t (x - 1),                              /* x */
+              int16_t (y + height), /* y */
```
Comment (TheHussky) [inline src/window.cpp]: format?

File: src/window.cpp [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        };
+  // Drawing the highlighting itself
+  xcb_poly_fill_rectangle (window::c_, this->id, desktop_pixmap::gc_, 4,
+                           rectangles);
+
+  // Changing color
+  uint32_t mask = XCB_GC_FOREGROUND;
+  values[0] = color;
+  xcb_change_gc (c_, desktop_pixmap::gc_, mask, &values);
+}
```
Comment (TheHussky) [inline src/window.cpp]: last time u asked for newline, maybe add one here?

**Review by TheHussky — APPROVED** [review]

**Review by TheHussky — APPROVED** [review]

**Pull request merged**